server/session/handler_item_stack_request.go

```diff
@@ -170,6 +176,62 @@ func (h *ItemStackRequestHandler) handleSwap(a *protocol.SwapStackRequestAction,
 	return nil
 }
 
+// handleCraft handles the Craft stack request action.
+func (h *ItemStackRequestHandler) handleCraft(recipeNetworkID uint32, auto bool, timesCrafted int, s *Session) error {
+	r, ok := s.recipeMapping[recipeNetworkID]
+	if !ok {
+		return fmt.Errorf("invalid recipe network id sent")
+	}
+
+	var expectedInputs []recipe.Item
+	var output item.Stack
+
+	switch r := r.(type) {
+	case recipe.ShapelessRecipe:
+		expectedInputs, output = r.Inputs, r.Output
+	case recipe.ShapedRecipe:
+		expectedInputs, output = r.Inputs, r.Output
+	default:
+		return fmt.Errorf("tried crafting an invalid recipe: %T", r)
+	}
+
+	if auto {
+		var newExpectedInputs []recipe.Item
+		for _, input := range expectedInputs {
+			newExpectedInputs = append(newExpectedInputs, recipe.Item{
+				Stack:    input.Grow((input.Count() * timesCrafted) - input.Count()),
+				AllTypes: input.AllTypes,
+			})
+		}
+
+		if !h.hasRequiredInventoryInputs(newExpectedInputs, s) {
+			return fmt.Errorf("tried crafting without required inventory inputs")
+		}
+
+		if err := h.removeInventoryInputs(newExpectedInputs, s); err != nil {
+			return err
+		}
+
+		output = output.Grow((output.Count() * timesCrafted) - output.Count())
```

Comment:
```suggestion
		output = output.Grow(output.Count() * (timesCrafted-1))
```

I'd do this else where too.

Reply:
Output counts aren't always one, for example, doors have an output of 3.

Reply:
Actually, nevermind, I read that wrong

Reply:
Changed.